### PR TITLE
ignore bucket-location flag if s4cmd detected

### PR DESCRIPTION
**Pull request by ithayer**

### CODE REVIEW

**Comment by ithayer** [comment]
@robertzk @kirillseva https://github.com/avantcredit/avant-analytics/issues/2684

**Comment by robertzk** [comment]
Thanks!

File: R/s3.get.R [hunk 1]
```diff
@@ -78,3 +81,20 @@ s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE
   }
   ans
 }
+
+## Given an s3cmd path and a bucket location, will construct a flag
+## argument for s3cmd.  If it looks like the s3cmd is actually
+## pointing to an s4cmd, return empty string as s4cmd doesn't
+## support bucket location.
+bucket_location_to_flag <- function(s3cmd_binary_path, bucket_location) {
+  if (grepl("s4cmd", s3cmd_binary_path)) {
```
Comment (peterhurford) [inline R/s3.get.R]: Maybe make this a helper if we'll need it in the future?

File: R/s3.get.R [hunk 2]
```diff
@@ -78,3 +81,20 @@ s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE
...
+    if (bucket_location != "US") {
```
Comment (peterhurford) [inline R/s3.get.R]: `identical(bucket_location, "US")` is preferred because of weird edge cases involving length >1 lists.

File: R/s3.get.R [hunk 3]
```diff
@@ -78,3 +81,20 @@ s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE
...
+        warning(paste0("Ignoring non-default bucket location ('",
+                       bucket_location,
+                       "') in s3mpi::s3.get since s4cmd was detected",
+                       "-- this might be a little slower but is safe to ignore."));
+    }
+    return("")
```
Comment (peterhurford) [inline R/s3.get.R]: R will always return the last evaluated thing by default, so explicit `return` isn't necessary except for early returns.

File: R/s3.get.R [hunk 4]
```diff
@@ -78,3 +81,20 @@ s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE
...
+## argument for s3cmd.  If it looks like the s3cmd is actually
+## pointing to an s4cmd, return empty string as s4cmd doesn't
+## support bucket location.
+bucket_location_to_flag <- function(s3cmd_binary_path, bucket_location) {
+  if (grepl("s4cmd", s3cmd_binary_path)) {
+    if (bucket_location != "US") {
+        warning(paste0("Ignoring non-default bucket location ('",
+                       bucket_location,
+                       "') in s3mpi::s3.get since s4cmd was detected",
+                       "-- this might be a little slower but is safe to ignore."));
```
Comment (peterhurford) [inline R/s3.get.R]: Nice user-friendly warning 👍

File: R/s3.get.R [hunk 5]
```diff
@@ -28,9 +28,12 @@ s3.get <- function (path, bucket.location = "US", verbose = FALSE, debug = FALSE
...
 
     ## Run the s3cmd tool to fetch the file from S3.
-    s3.cmd <- paste("get", paste0('"', path, '"'), x.serialized, paste("--bucket-location",
-      bucket.location), ifelse(verbose, "--verbose --progress",
-      "--no-progress"), ifelse(debug, "--debug", ""))
+    s3.cmd <- paste("get", paste0('"', path, '"'), x.serialized,
+                    bucket_location_to_flag(s3cmd(), bucket.location),
+                    ifelse(verbose,
+                           "--verbose --progress",
+                           "--no-progress"),
```
Comment (peterhurford) [inline R/s3.get.R]: `ifelse` isn't really meant as a ternary, it's meant as a vectorized case statement for length >1 vectors.

**Comment by peterhurford** [comment]
Nice PR. Made some post hoc comments but nothing that would prevent merging.

**Comment by ithayer** [comment]
good comments. note that i didn't add the ifelse, but will address other stuff in a new commit. is it best practice to update the package version?

**Comment by robertzk** [comment]
We generally update the package version for non-trivial changes so we can tag it with lockbox. Don't worry about the other comments - I added them quickly [here](https://github.com/robertzk/s3mpi/commit/b972ca1b5236774932d734c8ed968c615d87969a).

**Comment by ithayer** [comment]
thx! i note that you didn't up the package version, so for trivial/non-functional changes I assume we leave it

**Comment by robertzk** [comment]
For non-functional changes it's fine to leave as is; in this case I actually cheated and waited to tag 0.2.30 on top of that commit.

**Pull request merged**